✨ Add login

benfurber opened the pull request:
A barebones login feature. A user can login and be provided with a token that's sent with proceeding API requests. On logout, the token is cleared from the local cache.

It's a big PR, but not as much as it seems because of snapshots and typed-flow updates.

benfurber committed:
✨ Add empty Login screen

benfurber committed:
✨ Add login to navigation

App now loads to an authentication navigation stack. When a user is authenticated, it switches to the main stack.

benfurber committed:
✨ Add empty FormLogin component

benfurber committed:
✨ Add login button to FormLogin

benfurber committed:
♻️ Refactor handlePress into utility function

benfurber committed:
♻️ Refactor components to use utility handlePress

benfurber committed:
✨ Add form fields to FormLogin component

benfurber committed:
🚧 WIP - Add authentication redux flow

benfurber committed:
🐛 Fix double declaration of Goal in routes

benfurber committed:
🗑 Drop remaining redux code

These appeared because I rebased to an older branch.

benfurber committed:
🐛 Fix naming of uri

benfurber committed:
✨ Add FormLoginButton ready for Apollo

benfurber committed:
✨ Add loginMutation query to get token from API

benfurber committed:
♻️ Refactor FormLoginButton for Apollo

This included removing flow because of a conflict with the react-apollo definition.

benfurber committed:
✨ Update flow-typed defintions

benfurber committed:
✨ Add Apollo-link-state package

benfurber committed:
✨ Add FormLoginButton to FormLogin

benfurber committed:
♻️ Refactor props for button to come from props

benfurber committed:
♻️ Refactor client for authentication token

benfurber committed:
♻️ Refactor Login to auto-login if token present

benfurber committed:
🐛 Fix navigationMock call on test

benfurber committed:
🐛 Update failing App test

benfurber committed:
♻️ Refactor IconButton to use a DRY-er case switch

benfurber committed:
✨ Add Logout “screen”

benfurber committed:
✨ Add Logout screen to main navigation

benfurber committed:
🐛 Fix login flow

It actually works now.

* Moved token creation to FormLoginButton
* Passed props down the tree to FormLoginButton

benfurber committed:
♻️ Refactor authentication flow between screens

AuthLoading now decides if users should be sent to the login screen or the feed

benfurber committed:
🗑 Drop unused component import

benfurber committed:
✨ Link FormLogin inputs up

benfurber committed:
🗑 Drop props for LoginForm from Login

benfurber committed:
✨ Add value set for inputs

benfurber committed:
♻️ Refactor email and password to constant types

benfurber committed:
♻️ Refactor FormLogin test

- Remove defining props on render
- Update snapshot

tjmw commented on app/components/IconButton.js:
```diff
@@ -15,26 +15,26 @@ class IconButton extends Component<Props> {
-      return (
-        <Icon name="md-speedometer" style={[styles.icon, { color: tint }]} />
-      );
+  _selectIcon(name: string) {
```
This is kind of personal preference, but if you're interested in making this _actually_ private (as opposed to pseudo-private, indicated by the leading `_`) then you could pull this out to a function defined (but not exported) in this file. It's completely pure (i.e. no side-effects, doesn't care about any state) so it doesn't _need_ to be a method on the class.

tjmw commented on app/components/__tests__/FormLogin.test.js:
```diff
@@ -0,0 +1,16 @@
+
+describe("Loading", () => {
+  it("has a valid snapshot", () => {
+    const component = shallow(<FormLogin navigation={navigationMock} />);
```
Aha I'd not seen snapshot rendering done with Enzyme before. I've always used [`react-test-renderer`](https://www.npmjs.com/package/react-test-renderer). I guess one difference to be aware of is that this is shallow rendering (I don't think Enzyme supports deep rendering with React Native) so will only render one component layer deep, whereas `react-test-renderer` is deep rendering.

tjmw commented on app/components/__tests__/FormLoginButton.test.js:
```diff
@@ -0,0 +1,41 @@
+
+describe("Loading", () => {
+  it("has a valid snapshot", () => {
+    const mutateMock = () => jest.fn();
```
Nitpicky, but this doesn't need to be a jest fn, since you're not making any assertions on it. It could just be an empty anonymous function (`() => {}`). I think the only downside of using a jest mock is that it could prime the reader to look out for an assertion being made which never happens.

tjmw commented on app/routes.js:
```diff
@@ -41,6 +49,18 @@ const tabConfig = {
+  AuthLoading: {
+    screen: AuthLoading,
+  },
+});
```
I _think_ this can be simplified to:

```js
const AuthLoadingNavigator = createStackNavigator({
   AuthLoading,
 });
```

benfurber replied:
Ugh. Of course, thanks.

tjmw commented on app/routes.js:
```diff
@@ -41,6 +49,18 @@ const tabConfig = {
+  Login: {
+    screen: Login,
+  },
+});
```
Same comment as above with `AuthLoading`.

tjmw commented on app/screens/AuthLoading.js:
```diff
@@ -0,0 +1,39 @@
+    this.asyncLogin();
+  }
+
+  asyncLogin = async () => {
```
It doesn't _seem_ like there's a reason to use this style of method definition in this case. How about the standard:

```js
async asyncLogin() {
  ...
}
```

Out of interest, is prepending async methods/functions with `async` a pattern you're experimenting with? Or is this a one off?

tjmw commented on flow-typed/npm/apollo-link-context_vx.x.x.js:
```diff
@@ -0,0 +1,46 @@
+// flow-typed signature: d2e9c275593d777b12757d2dbe083d03
```
I wonder if it would be worth hiding these by default, like we do with snapshots?

tjmw reviewed (commented):
This is looking good to me! I'd be tempted to try and source a review from someone with more Apollo/GraphQL experience than me.

Is there logic anywhere to handle the API returning a `401 Unauthorized` in the case of an expired/bad token? Or it that something magically handled for us somewhere?

benfurber committed:
🗑 Drop unneeded onPress prop

benfurber committed:
🗑 Drop unused jest mock

benfurber committed:
♻️ Refactor navigators to be less verbose

benfurber committed:
♻️ Refactor asyncLogin()

- Use standard async class definition
- Define routeName more clearly

benfurber committed:
♻️ Refactor Navigation type

The current type-floe definition isn’t written very well so the getParam and navigate overwrites are needed to not get any errors.

benfurber committed:
🐛 Fix list prop

benfurber committed:
🔧 Hide flow-type folder from PR display
